File: src/resources/status.ts
import type { HttpClient } from '../http.js';
import type { LiveStatusEvent } from '../types/misc.js';

export type StatusEventHandler = (event: LiveStatusEvent) => void;

export interface LiveStatusConnection {
  close(): void;
}

export class StatusResource {
  constructor(private http: HttpClient) {}

  async connect(onEvent: StatusEventHandler): Promise<LiveStatusConnection> {
    const response = await this.http.raw('/v1/status/live', {
      headers: { Accept: 'text/event-stream' },
    });

    if (!response.body) {
      throw new Error('SSE: No response body');
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let aborted = false;

    const processStream = async () => {
      while (!aborted) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() ?? '';

        let eventType = '';
        let eventData = '';

        for (const line of lines) {
          if (line.startsWith('event:')) {
            eventType = line.slice(6).trim();
          } else if (line.startsWith('data:')) {
            eventData = line.slice(5).trim();
          } else if (line === '') {
            if (eventType && eventData) {
              try {
                const data = JSON.parse(eventData);
                onEvent({ type: eventType as LiveStatusEvent['type'], data });
              } catch {
                onEvent({ type: eventType as LiveStatusEvent['type'], data: eventData });
              }
            }
            eventType = '';
            eventData = '';
          }
        }
      }
    };

    processStream().catch(() => {});

    return {
      close() {
        aborted = true;
        reader.cancel().catch(() => {});
      },
    };
  }
}

File: packages/python/invariance/resources/status.py
from __future__ import annotations

from typing import Any, Callable

from ..http_client import HttpClient
from ..types import LiveStatusEvent, LiveStatusSnapshot


class LiveStatusConnection:
    def __init__(self, close_fn: Callable[[], None]) -> None:
        self._close_fn = close_fn

    def close(self) -> None:
        self._close_fn()


class StatusResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def snapshot(self) -> LiveStatusSnapshot:
        return await self._http.get("/v1/status/live")

    async def connect(
        self, on_event: Callable[[LiveStatusEvent], None]
    ) -> LiveStatusConnection:
        import asyncio
        import json

        response = await self._http.raw(
            "/v1/status/live", headers={"Accept": "text/event-stream"}
        )

        aborted = False

        async def process_stream() -> None:
            nonlocal aborted
            buffer = ""
            event_type = ""
            event_data = ""

            async for chunk in response.aiter_text():
                if aborted:
                    break
                buffer += chunk
                lines = buffer.split("\n")
                buffer = lines.pop()

                for line in lines:
                    if line.startswith("event:"):
                        event_type = line[6:].strip()
                    elif line.startswith("data:"):
                        event_data = line[5:].strip()
                    elif line == "":
                        if event_type and event_data:
                            try:
                                parsed = json.loads(event_data)
                                on_event(parsed)
                            except Exception:
                                pass
                        event_type = ""
                        event_data = ""

        task = asyncio.get_running_loop().create_task(process_stream())

        def close() -> None:
            nonlocal aborted
            aborted = True
            task.cancel()

        return LiveStatusConnection(close)
